lib/domain/usecase/search/search_interactor.dart

```diff
@@ -0,0 +1,79 @@
+import 'package:fluffychat/domain/model/extensions/contact/presentation_contact_list_extension.dart';
+import 'package:fluffychat/domain/repository/contact_repository.dart';
+import 'package:fluffychat/presentation/extensions/room_extension.dart';
+import 'package:fluffychat/presentation/extensions/room_list_extension.dart';
+import 'package:fluffychat/presentation/model/presentation_search.dart';
+import 'package:fluffychat/utils/matrix_sdk_extensions/client_stories_extension.dart';
+import 'package:fluffychat/utils/matrix_sdk_extensions/matrix_locals.dart';
+import 'package:fluffychat/widgets/matrix.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+import 'package:flutter/cupertino.dart';
+import 'package:matrix/matrix.dart';
+
+class SearchContactsAndRecentChatInteractor {
+  final ContactRepository contactRepository = getIt.get<ContactRepository>();
+
+  SearchContactsAndRecentChatInteractor();
+
+  Stream<Either<Failure, GetContactAndRecentChatSuccess>> execute({
+    required BuildContext context,
+    required String keyword,
+    int? limitContacts,
+    int? limitRecentChats,
+    bool enableSearch = false
+  }) async* {
+    try {
+      if (enableSearch) {
+        final recentChat = await _searchRecentChat(context, keyword);
+        final contacts = await contactRepository.searchContact(query: ContactQuery(keyword: keyword), limit: limitContacts);
+
+        final presentationSearches = _comparePresentationSearches(
+          recentChat.toPresentationSearchList(context),
+          contacts.expand((contact) => contact.toPresentationContacts()).toList().toPresentationSearchList()
+        );
+        yield Right(GetContactAndRecentChatSuccess(presentationSearches: presentationSearches));
+      } else {
+        final rooms = await _getRecentChat(context, limitRecentChats: limitRecentChats);
+        yield Right(GetContactAndRecentChatSuccess(presentationSearches: rooms.toPresentationSearchList(context)));
+      }
+    } catch (e) {
+      yield Left(GetContactAndRecentChatFailed(exception: e));
+    }
+  }
+
+  Future<List<Room>> _getRecentChat(BuildContext context, {int? limitRecentChats}) async {
+    return Matrix.of(context).client.rooms.where((room) => !room.isSpace && !room.isStoryRoom)
+      .where((room) => room.isShowInChatList())
+      .take(limitRecentChats ?? 0)
+      .toList();
+  }
+
+  Future<List<Room>> _searchRecentChat(BuildContext context, String keyword) async {
+    return Matrix.of(context).client.rooms.where((room) => !room.isSpace && !room.isStoryRoom)
+      .where((room) => room.isShowInChatList())
+      .where((room) =>
+        room.getLocalizedDisplayname(MatrixLocals(L10n.of(context)!))
+        .toLowerCase()
+        .contains(keyword.toLowerCase())
+    ).toList();
+  }
+
+  List<PresentationSearch> _comparePresentationSearches(List<PresentationSearch> recentChat, List<PresentationSearch> contacts) {
+    final isDuplicateElement = contacts.where((contact) {
```

Comment:
WDYT if use it like this.
`var mergedList = [...listX, ...listY, ...listZ].toSet(); `

Reply:
in your code I tried it and then result:

```
ListX = [e1, e2, e3]
ListY = [e4, e5, e6]

Noted: `e1.id == e4.id`

var mergedList = [...listX, ...listY].toSet().toList(); 

=> mergedList =  [e1, e2, e3, e4, e5, e6]
```

My code 

```
ListX = [e1, e2, e3]
ListY = [e4, e5, e6]

Noted: `e1.id == e4.id`

varr isDuplicateElement = ....(my code)

=> mergedList =  [e1, e2, e3, e5, e6]
```
IMO: I just want to create a list of duplicated elements and remove it from `mergedList`